### PR TITLE
519 Add Basic GAN Workflow

### PR DESCRIPTION
**Pull request by gagandaroach**
Progress towards #531 completion.

### Description
With this PR I introduce
* Monai GanTrainer engine.
  * Training algorithm based on Goodfellow 2014 GAN paper
* Workflow example showing training GAN to generate Hands with MedNIST dataset.
  * trainer and evaluator script
  * in `monai.examples.synthesis`
  * based on Monai example MedNIST Gan Jupyter notebook.
* `create_run_dir` function in `monai.utils.misc`
  * IMO a controversial addition, this function creates a timestamped output directory for saving a current run's data.

### Status
**In Review**

- [x] Add ipynb example
- [x] Update example readme

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### Comments
* Right now the GAN workflow test is a smoke test, checking if the trainer completes 2 epochs with input without errors.
* In future for Monai GAN functionality,
     * I want to make a handler to save generator output with fixed test latents every n epochs.
     * Will need to add a key GAN train metric in Monai e.g. FID https://arxiv.org/abs/1706.08500

### CODE REVIEW

**Comment by gagandaroach** [comment]
@Nic-Ma @wyli This is my first milestone for #531 . I am not seeing how to add a required reviewer to the PR.

**Comment by Nic-Ma** [comment]
Hi @gagandaroach ,

Thanks for your quick PR, as the error message said:
```
monai/utils/misc.py:222: error: unused 'type: ignore' comment
monai/utils/misc.py:226: error: Module "datetime" is not valid as a type
monai/engines/trainer.py:280: error: Dict entry 1 has incompatible type "str": "int"; expected "str": "Tensor"
Found 3 errors in 2 files (checked 313 source files)
```
Could you please help update them?

Thanks.

**Comment by gagandaroach** [comment]
Hello @Nic-Ma 
I was able to resolve the CI errors.

**Comment by gagandaroach** [comment]
I refractored and added some unit tests. 

I find CI to fail because flake8 and black do not agree on the position of a space.
```py
    train_transforms = Compose(
        [LoadNiftid(keys=["reals"]), ScaleIntensityd(keys=["reals"]), ToTensord(keys=["reals"]),X]
    )
```
flake8 wants a space on the X. black wants no space at the X. Not sure what to do.

**Comment by gagandaroach** [comment]
I am refactoring the code to handle a dataloader and dataset for generator.

**Comment by gagandaroach** [comment]
> I am refactoring the code to handle a dataloader and dataset for generator.

I decided against this for now. The D_data_loader does not return indices, so it is not easy to correlate with values coming from G_data_loader. I think this simpler workflow with one D_data_loader is good.

I think for RGGAN i can accomplish dynamic generator input with a modified g_inferer and make_latent.

File: monai/engines/trainer.py [hunk 1]
```diff
@@ -145,3 +146,145 @@ def _iteration(self, engine: Engine, batchdata: Union[Dict, Sequence]) -> Dict[s
...
+        g_inferer: Inferer = SimpleInferer(),
+        d_inferer: Inferer = SimpleInferer(),
+        d_train_steps: int = 1,
+        latent_shape: int = 64,
+        prepare_batch: Callable = default_prepare_batch,
+        make_latent: Callable = make_rand_latent_code,
+        iteration_update: Optional[Callable] = None,
+        amp: bool = True,
+        post_transform: Optional[Transform] = None,
+        key_train_metric: Optional[Dict[str, Metric]] = None,
```
Comment (wyli) [inline monai/engines/trainer.py:213]: do we need to expose this `key_train_metric` option?

File: monai/engines/trainer.py [hunk 2]
```diff
@@ -145,3 +146,145 @@ def _iteration(self, engine: Engine, batchdata: Union[Dict, Sequence]) -> Dict[s
         self.optimizer.step()
 
         return {Keys.IMAGE: inputs, Keys.LABEL: targets, Keys.PRED: predictions, Keys.LOSS: loss.item()}
+
+
+class GanTrainer(Trainer):
+    """
+    Standard generative adversarial network training, inherits from trainer and Workflow.
```
[the remark this answers is not in the document]
Comment (gagandaroach) [reply monai/engines/trainer.py:153]: I agree, there are many GAN training loops. Here I imply simple in the sense of one latent input into the generator.

I think another trainer workflow class should be made for GANs that provide supplementary input into the generator like MC-GAN or Dlmed's RGGAN. I imagine one prepare_batch func processing the dataloader batch for both G and D.

File: monai/engines/trainer.py (hunk 1 above)
Comment (gagandaroach) [reply monai/engines/trainer.py:213]: I think for future implementations yes. Like StyleGAN or BigGAN, many computer vision GANs use distribution comparison calculations every n epoch's as a key training metric (FID, PPL, IS)

Also since some GAN networks collapse at scale, this key_train_metric could be used with early stopping Monai handlers.

File: monai/utils/misc.py [hunk 3]
```diff
@@ -217,3 +220,34 @@ def set_determinism(
         torch.backends.cudnn.benchmark = False
     else:
         torch.backends.cudnn.deterministic = False
+
+
+def create_run_dir(
```
[the remark this answers is not in the document]
Comment (gagandaroach) [reply monai/utils/misc.py:225]: Okay I will refractor this.

**Comment by gagandaroach** [comment]
> thanks, would be great to have a jupyter notebook or colab to demonstrate some working examples

Should I modify the Monai MedNIST GAN notebook to exemplify this trainer or introduce a new notebook?

**Comment by wyli** [comment]
> > thanks, would be great to have a jupyter notebook or colab to demonstrate some working examples
> 
> Should I modify the Monai MedNIST GAN notebook to exemplify this trainer or introduce a new notebook?

please have a new notebook, preferably a colab one, if it's a lightweight demo

File: monai/engines/trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gagandaroach) [reply monai/engines/trainer.py]: I thought through the problem more. I defined a g_prepare_batch callback to handle auxiliary input into the generator. Now the proposed GanTrainer handles "standard" latent input and dynamic input.

**Comment by gagandaroach** [comment]
I added a jupyter notebook example but it needs some help. I was not sure how to import this new code into the notebook, so right now running it in Google colab crashes. Local execution works fine though.

**Comment by gagandaroach** [comment]
Please re-review my PR. I added a jupyter notebook and a docs entry.
